### PR TITLE
Bug 1254599 app state

### PR DESCRIPTION
**Pull request by fluffyemily**
Create an AppState enum that we can use to pass the current state of the app into various objects to allow them to make appropriate decisions without the need to reference the BrowserViewController

Questions here are:

We want to refresh the menu when the AppState changes. What is the best way of going about this?

1. If MenuConfiguration was a class rather than a struct, could we use KVO to observe the app state value changing and react to that. If that is the case would we need to keep a handle on the menu configuration in the calling object (BVC, TabTrayController etc) in order to keep this updated?

2. If we store app state outside of the menuConfig but inside the menu, (i.e. pass in a menu config & an app state), keep a handle on the MenuVC inside the calling object and then update that app state directly when it changes.

3. Fire off a NSNotification when any app state value changes which causes any referring objects to fetch a new instance of the app state (not really sure how this would work. would need more thought)

### CODE REVIEW

**Comment by sleroux** [comment]
I have mixed feelings over the AppState design/implementation. Some issues I see:

1. I feel like we're introducing another layer of caching to contend with and abstracting away from the source of truth of when we are in private mode, when the page is bookmarked, etc. I think it might make sense to consider folding these aspects into `Browser` but I'm not sure how to go about that atm. 

2. Retaining the state as static vars on the AppState enum isn't much different than using globals. Although in a sense we do want some global notion of state, I feel like we can compartmentalize it better and place it alongside existing models like `Browser`. The other downside of having static vars is it becomes more difficult to mock and test with since we won't be able to inject state.

3. I'm not sure of the value of having the data live inside the enum. Most places we access the value directly to the static var in the enum which weakens the relationship between the specific enum and it's associated values. I think extracting the data into another abstraction and keeping the enum to just describe the various pages in the app would be cleaner.

This problem is definitely a PITA and I'm not sure of the best way to move forward. I think we should bring in the team and brainstorm around it since it seems like a pretty core aspect of the app that we'll want to extend on in the future. Maybe put it on the agenda for tomorrow's team meeting?

File: Client/Application/AppState.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+enum AppState {
+    case Browser(currentURL: NSURL, isBookmarked: Bool, isDesktopSite: Bool, hasAccount: Bool, isPrivate: Bool)
+    case HomePanels(selectedPanelIndex: Int, isPrivate: Bool)
```
Comment (jhugman) [inline Client/Application/AppState.swift]: Using an `enum` for state seems to be stretching the meaning of an enumerated type.

Also: can we move away from calling a tab a browser.

File: Client/Application/AppState.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    private static func getLoadingState() -> AppState {
+        return .Loading
+    }
+
+    static var url: NSURL?
+    static var bookmarked: Bool = false
+    static var isDesktop: Bool = false
+    static var hasAccount: Bool = false
+    static var isPrivate: Bool = false
+    static var homePanelState: Int?
```
Comment (jhugman) [inline Client/Application/AppState.swift]: Global state smell.

**Comment by jhugman** [comment]
I'm finding it hard to like the `AppState` as an `enum`. Using it to store a URL seems to be going against the spirit of enumerability.

I am most concerned about the statics.

It feels like this is yet another source of truth that needs to be kept upto data, and impossible to unit test or re-use.

I think I'd suggest an `AppState` `protocol` and an implementation that bridges/adapts the existing objects.

**Pull request closed without merging**